node_groups.tf

```diff
@@ -1,112 +1,29 @@
-}
-
-resource "aws_iam_role_policy_attachment" "node_groups_AmazonEC2ContainerRegistryReadOnly" {
-  count      = var.create_eks && local.worker_group_managed_node_group_count > 0 ? 1 : 0
-  policy_arn = "arn:aws:iam::aws:policy/AmazonEC2ContainerRegistryReadOnly"
-  role       = aws_iam_role.node_groups[0].name
-}
-
-resource "aws_iam_role_policy_attachment" "node_groups_additional_policies" {
-  for_each = var.create_eks && local.worker_group_managed_node_group_count > 0 ? toset(var.workers_additional_policies) : []
-
-  role       = aws_iam_role.node_groups[0].name
-  policy_arn = each.key
-}
-
-resource "aws_iam_role_policy_attachment" "node_groups_autoscaling" {
-  count      = var.create_eks && var.manage_worker_autoscaling_policy && var.attach_worker_autoscaling_policy && local.worker_group_managed_node_group_count > 0 ? 1 : 0
-  policy_arn = aws_iam_policy.node_groups_autoscaling[0].arn
-  role       = aws_iam_role.node_groups[0].name
-}
-
-resource "aws_iam_policy" "node_groups_autoscaling" {
-  count       = var.create_eks && var.manage_worker_autoscaling_policy && local.worker_group_managed_node_group_count > 0 ? 1 : 0
-  name_prefix = "eks-worker-autoscaling-${aws_eks_cluster.this[0].name}"
-  description = "EKS worker node autoscaling policy for cluster ${aws_eks_cluster.this[0].name}"
-  policy      = data.aws_iam_policy_document.worker_autoscaling[0].json
-  path        = var.iam_path
-}
-
-resource "random_pet" "node_groups" {
-  for_each = var.create_eks ? local.node_groups : {}
-
-  separator = "-"
-  length    = 2
-
-  keepers = {
-    instance_type = lookup(each.value, "instance_type", local.workers_group_defaults["instance_type"])
-
-    ec2_ssh_key = lookup(each.value, "key_name", local.workers_group_defaults["key_name"])
-
-    source_security_group_ids = join("-", compact(
-      lookup(each.value, "source_security_group_ids", local.workers_group_defaults["source_security_group_id"]
-    )))
-
-    node_group_name = join("-", [var.cluster_name, each.value["name"]])
+# Hack to ensure ordering of resource creation. Do not create node_groups
+# before other resources are ready. Removes race conditions
+data "null_data_source" "node_groups" {
+  count = var.create_eks ? 1 : 0
+
+  inputs = {
+    cluster_name = var.cluster_name
+
+    # Ensure these resources are created before "unlocking" the data source.
+    # `depends_on` causes a refresh on every run so is useless here.
+    # [Re]creating or removing these resources will trigger recreation of Node Group resources
+    aws_auth         = coalescelist(kubernetes_config_map.aws_auth[*].id, [""])[0]
+    role_NodePolicy  = coalescelist(aws_iam_role_policy_attachment.workers_AmazonEKSWorkerNodePolicy[*].id, [""])[0]
+    role_CNI_Policy  = coalescelist(aws_iam_role_policy_attachment.workers_AmazonEKS_CNI_Policy[*].id, [""])[0]
+    role_Container   = coalescelist(aws_iam_role_policy_attachment.workers_AmazonEC2ContainerRegistryReadOnly[*].id, [""])[0]
```

Comment:
Just got a thought, should we create distincts IAM policies for each submodules ? This will make them totally self contained from each others.

@max-rocket-internet @dpiddockcmp any thoughts ?

Reply:
I don't think so by default. Everything can use the same single worker role between all workers. AWS seems to be pushing towards this role being as basic as possible with pods using OIDC where necessary. For example see the recently updated [blog post about ALB ingress](https://aws.amazon.com/blogs/opensource/kubernetes-ingress-aws-alb-ingress-controller/)

If users really need different IAM roles per worker group they can always override in one of the maps.

Reply:
In a near future we will split the terraform-aws-eks module into multiple submodules. Probably like this:
- `modules/node_groups` (the current PR)
- `modules/fargate_profile`
- `modules/workers`
- `modules/cluster` ??? (this can stay in the root module)

The root module will probably only create the cluster. So I was thinking to move `
aws_iam_role_policy_attachment.*` into each submodules and with their own iam_role. In that way, user can provide aditionnal policy to a particular module. I see this already in cloudpose modules [aws-eks-workers](https://github.com/cloudposse/terraform-aws-eks-workers/blob/master/main.tf#L38-L73), [aws-eks-node-group](https://github.com/cloudposse/terraform-aws-eks-node-group/blob/master/main.tf#L35-L64), etc.

In other words, my thoughts were: should we split IAM role and profile per submodule in the near future or not ? If yes, should we start now with the node_groups submodule ?

Reply:
TBH I am with @barryib on this. Let's aim for as much separation as possible.

> In a near future we will split the terraform-aws-eks module into multiple submodules

Well, let's see how PR/module goes first 😅

Reply:
Ahh but then the IAM roles need to be passed back out of the NG module to be used in `aws-auth` configmap, which might create a circular dependency?

Reply:
> Ahh but then the IAM roles need to be passed back out of the NG module to be used in aws-auth configmap, which might create a circular dependency?

Hummm... I don't think so. Terraform will handle the dependency graph. The aws-auth templating will only wait for the IAM role creation and not for the whole module.

Reply:
Why do we need all that code duplication?

If you want to use cloudposse modules why not use them instead of slowly migrating this one to being a copy? It makes slightly more sense for them as they're totally separate module repos.

This module was a one shot for setting up everything EKS. From that point it makes more sense to only have the one worker role.

Reply:
@dpiddockcmp The idea here is not mimicking those modules, but discussing how we should migrate slowly from on main modules to a multiple submodules with all his pros and cons.

> It makes slightly more sense for them as they're totally separate module repos.

Make sense.

Let merge this PR and get feedbacks from users. We'll improve it if needed.

> Well, let's see how PR/module goes first 😅

@max-rocket-internet I know we are not in the hurry, but if we start splitting this module, we have to continue with other features (fargate and actual workers).

Thanks all for your time for discussions and work.

Reply:
> Let merge this PR and get feedbacks from users. We'll improve it if needed.

Agreed!